lib/hanami.rb

```diff
@@ -8,4 +10,64 @@
 module Hanami
   require "hanami/version"
   require "hanami/frameworks"
+  require "hanami/container"
+  require "hanami/application"
+
+  @_mutex = Mutex.new
+
+  def self.application
+    @_mutex.synchronize do
+      raise "Hanami application not configured" unless defined?(@_application)
+
+      @_application
+    end
+  end
+
+  def self.application=(app)
+    @_mutex.synchronize do
+      raise "Hanami application already configured" if defined?(@_application)
+
+      @_application = app unless app.name.nil?
+    end
+  end
+
+  def self.app
```

Comment:
I wonder if we want a different name for this, since the pairing of `application`/`app` may be a little confusing?

Reply:
@timriley This **is** the only instance of `Hanami::Application` subclass in the process. So it's **the** app.

```ruby
Hanami.app 
# => #<Soundeck::Application:0x00007ffee940eee0 ...>
```

Reply:
Ah, so for a booted application, `Hanami.application` is the application's _class_ (via the `Hanami::Application.inherited` hook) and `Hanami.app` is the one _instance_ of that application class? Is that right? i.e.

```ruby
Hanami.application
# => Soundeck::Application

Hanami.app
# => #<Soundeck::Application:0x00007ffee940eee0 ...>
```

I still feel like that's a pretty confusing pairing of names. "app" is just shorthand for "application", they're synonyms. It feels a little odd that we use them both here, holding 2 different (though obviously related) things.

Reply:
@timriley Which name do you suggest? 😃 Please remember that `Hanami.app` is short and memorable, and it's used by public facing integration specs.

Eg.

```ruby
require "features_helper"

RSpec.feature "Signup" do
  let(:app) { Hanami.app } # for capybara and rack-test
end
```

Reply:
@jodosha What about `project` for the container application? You used "project" four times in the main message to refer to this :) It's not quite as memorible or short as `app`, but it's not hard to remember either. I agree with @timriley that using `app` and `application` as different things would be confusing.

Reply:
I'm not sure to if keeping project is a good idea. The intention is to remove the word "project" from the terminology. See my YT video.

Reply:
Maybe we could stick with app in both methods but disambiguate between the two with a suffix, e.g. `#app` and `#app_class`?

Reply:
Given that `app` returns an instance of the application, I would say a natural name for the other method should be `app_class`. I would also say this should be a private API anyway.

Reply:
I can understand some of the confusion between "Hanami project" referring to this whole effort, but I still think "project" is a natural and obvious way to talk about what the "umbrella" app is.

Will `apps/` be renamed to `slices/`? They're still Rack-compatible apps. While I like the naming of "slices" and it seems appropriate, I feel like this is re-inventing the wheel and causes more confusion than it's worth. `Hanami.app` (or `Hanami.application`) being a different thing from `Hanami::App` is confusing. If we want to use "slice", we should go all the way and rename `Hanami::App` to `Hanami::Slice`. I don't think we should do that, but we should be consistent if we're changing the terminology.

Django uses the `project`/ `application` distinction in the same way we have been: https://medium.com/@timmykko/a-quick-glance-of-django-for-beginners-688bc6630fab
> Any Django _project_ will have at least one Django _app_. Yes, I know, it’s a little confusing. A Django _project_ encompasses the application and all its components, while a Django _app_ is a sub-container of the application with its own functionality that, in theory, may be reusable in another application without much modification.


(Also, for others, [here's the YouTube video](https://www.youtube.com/watch?v=mQCTNswIhl8), since it hasn't been linked in this PR yet)

Reply:
@jodosha Thoughts on the above? ☝️  (Don't mean to bother you, I just want to make sure you saw it since I didn't mention you in the comment.) 🌸